### DISCUSSION TITLE
Access Token - roles

### DISCUSSION THREAD

**Discussion by vitalybrandes** [zitadel/zitadel#7594] [Q&A]
Hi ,
I need some help with access token, as looks like i am missing something.
I am receiving access token from zitadel with claims.
One of the claims is "urn:zitadel:iam:org:project:246506423876379555:roles"
and additional one is "urn:zitadel:iam:org:project:roles"
the each item include role, id and domain.
 "admin:admin": {
      "246505807364084345": "container.zitadel.cloud"
    },

As i have allot of roles ,this make my token to be really big.
how can i clean the access token and send only permission i created ""admin:admin" as example?
As well do not send both "urn:zitadel:iam:org:project:246506423876379555:roles" and "urn:zitadel:iam:org:project:roles" claims in the token.

thanks

**Comment by hifabienne** [comment]
Hi @vitalybrandes 
At the moment it is not possible to define which of those claims should be returned. 
One of them is the deprecated one and the other the one for the future, we tried to avoid breaking changes with that.
@livio-a can probably tell more

**Reply by vitalybrandes** [reply]
Thank you @hifabienne for quick response,

Regarding the deprecated claim, maybe its possible to add some checkbox to UI where Admin can decide what type of claim he want to get? Similar to this
![image](https://github.com/zitadel/zitadel/assets/61165619/db52d47b-c42b-4492-a911-e6d115b9551b)


For my case, i am building large scare micro-service application where permission are more than critical.
In some cases my user have 40 + permissions in summary.
I tried to use **Zitadel Action** to drop the default roles list and create my own, but also with no success. do you think this can be solution for me?

_Forgot to mention that i am using sass zitadel product and i am interested to create paid subscription once we will understand that solution is right for us.

**Reply by hifabienne** [reply]
A workaround at the moment could be that you use opaque tokens and call the userinfo endpoints afterwards.
In that case your tokens do not het to big.

**Comment by vitalybrandes** [comment]
@livio-a  
Can you help me somehow with this case?
Otherwise is there any support email?

**Reply by livio-a** [reply]
I see you problem, but unfortunately there's no possibility other than mentioned by Fabi just before: https://github.com/zitadel/zitadel/discussions/7594#discussioncomment-8874465
If you have 40+ roles, that could potentially be even an issue with "only" one of those lists.
We can think about providing an option or remove the `urn:zitadel:iam:org:project:roles` at one point, but at least the latter will take time, since it would introduce a breaking change for existing customers.
The best current way is to use an opaque token (you still get the JWT id_token for your app) and use the introspection / userinfo endpoint in your apis, where needed. To compensate latency, you could cache the response.

**Reply by vitalybrandes** [reply]
@livio-a  what about Zitadel Action , can i some how use this engine to create new list of permissions claims instead of roles?

**Reply by livio-a** [reply]
Hey @vitalybrandes yes you can add a custom claim to be able to have your own structure:
https://zitadel.com/docs/guides/integrate/retrieve-user-roles#customize-roles-using-actions

**Reply by vitalybrandes** [reply]
Can you please support me with that,

- What type of action i need to create to be able create my permissions claims? 
- What type of claim in my app i need to call?
- Can it be for both, access token and Id_Token?



thank you so much!!!

**Comment by vitalybrandes** [comment]
@livio-a 
I will appreciate it if you can check it on your side.
See following:
I setup "permissions" action in action page that include code:
```

function flatRoles(ctx, api) {
  if (ctx.v1.user.grants == undefined || ctx.v1.user.grants.count == 0) {
    return;
  }
  ctx.v1.user.grants.grants.forEach(claim => {
    claim.roles.forEach(role => {
api.v1.claims.setClaim('permission', role)
    })
  })
}
```
![image](https://github.com/zitadel/zitadel/assets/61165619/5d7fb060-3105-459d-8fdc-8bc65b11b4bb)

In my UI side i use this scope:
    _"Scope": "openid permission urn:zitadel:iam:action:permission:log",_


Getting resposne error:
{"error":"server_error","error_description":"Errors.Internal"}

Request URL:
https://1nstance-01-u1bal1.zitadel.cloud/oauth/v2/token
Request Method: POST
Status Code: 500 Internal Server Error

In the Zitadel events page i see following:

Access Token created | 246507009652876239User | 246505807364084063 | zitadel | 1248 | Mon 25. Mar, 12:40:23

{
  "applicationId": "25010635557112079@mobile_spa",
  "audience": [
    "2501063255797112079@mobile_spa",
    "251458844219830674@mobile_spa",
    "250108768566451471"
  ],
  "expiration": "2024-03-25T22:40:23.34924907Z",
  **"scopes": [
    "openid",
    "urn:zitadel:iam:org:project:role:create:mobile",
    "urn:zitadel:iam:org:project:role:delete:mobile",
    "urn:zitadel:iam:org:project:role:manage:mobile",
    "urn:zitadel:iam:org:project:role:readall:mobile",
    "urn:zitadel:iam:org:project:role:read:mobile",
    "urn:zitadel:iam:org:project:role:rent:mobile",
    "urn:zitadel:iam:org:project:role:reserve:mobile",
    "urn:zitadel:iam:org:project:role:return:mobile",
    "urn:zitadel:iam:org:project:role:update:mobile"
  ],**
  "tokenId": "259824084529602823",
  "userAgentId": "256049472551272956"
}

Please please , what i am doing wrong, otherwise is it bug?
Test it for both, Access_token and ID_token creation.

*P.S I am using your saas solution, now on free subscription before purchase.

thanks

**Reply by vitalybrandes** [reply]
@livio-a  or @hifabienne  can you please assist me with that?

**Reply by livio-a** [reply]
Hey @vitalybrandes 

With your current solution, you'd overwrite the claim with every role. ZITADEL does not allow that.
From what i see (list of roles without project or organization ID), the following should work for you:
```
let permissions = [];
  ctx.v1.user.grants.grants.forEach(claim => {
    claim.roles.forEach(role => {
        permissions.push(role)  
    })
  })

  api.v1.claims.setClaim('permissions', permissions)
```

**Reply by vitalybrandes** [reply]
@livio-a 
I don't know why , i am receiving 500 error.
Using code you share above.
![image](https://github.com/zitadel/zitadel/assets/61165619/b10ef94b-2c8b-4322-bb0c-3cc2c4c082ca)
![image](https://github.com/zitadel/zitadel/assets/61165619/81cae342-0e2f-4a3c-8ab1-1ca86444eec9)

Did you have any chance to check it on your side?

**Comment by vitalybrandes** [comment]
@livio-a  @hifabienne  Can you please assist with my question?
